File: app/src/main/java/jp/wildtree/android/apps/hhsadvrev/ZWord.java
package jp.wildtree.android.apps.hhsadvrev;

public class ZWord {
	public String word;
	public int id;
	
	public static final int INVALID_WORD = 0;
	
	public ZWord(byte[] b)
	{
		word = "";
		id   = -1;
		for (int i = 0 ; i < 4 ; i++)
		{
			int z = (b[i] & 0xff);
			if (z == 0) break;
			word += String.format("%c", (byte)(z - 1));
		}
		id = b[4];
	}
	
	public boolean match(String v)
	{
		String z = v + "    ";
		z = z.substring(0, 4).toUpperCase(); // normalie
		return z.equalsIgnoreCase(word);
	}
}

File: app/src/main/java/jp/wildtree/android/apps/hhsadvrev/ZWord.kt
package jp.wildtree.android.apps.hhsadvrev

import java.util.Locale

class ZWord(b: ByteArray) {
    var id: Int = -1
        private set
    var word: String = ""
        private set

    init {
        for (i in 0..3) {
            val z = b[i].toUByte().toInt()
            if (z == 0) break
            word += String.format("%c", (z - 1).toByte())
        }
        id = b[4].toInt()
    }

    fun match(v: String?): Boolean {
        var z = "$v    "
        z = z.substring(0, 4).uppercase(Locale.getDefault()) // normalize
        return z.equals(word, ignoreCase = true)
    }

    companion object {
        const val INVALID_WORD: Int = 0
    }
}
